condor_git_config.py

```diff
@@ -91,127 +98,174 @@ class ConfigCache(object):
+                raise RuntimeError(
+                    "config cache %r used for conflicting hooks" % self.cache_path
+                )
+            return meta_data
+
+    def _write_meta(self):
+        if self._config_meta is None:
+            return
+        with open(self._meta_file, "w") as meta_stream:
+            json.dump(self._config_meta, meta_stream)
+
+    def abspath(self, *rel_paths: Union[str, Path]) -> Path:
+        return self._work_path.joinpath(*rel_paths)
+
+    def repo_path(self, *rel_paths: Union[str, Path]) -> Path:
+        return self.abspath("repo", *rel_paths)
 
     def __enter__(self):
         self._cache_lock.acquire()
+        self._config_meta = self._read_meta()
+        self._refresh()
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        self._write_meta()
+        self._config_meta = None
         self._cache_lock.release()
         return False
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[Path]:
+        assert self._config_meta is not None
         # avoid duplicates from links
         seen = set()
         repo_path = self.repo_path()
         for dir_path, dir_names, file_names in os.walk(repo_path):
-            if '.git' in dir_names:
-                dir_names.remove('.git')
-            dir_names[:] = sorted(dir_names)
+            try:
+                dir_names.remove(".git")
+            except ValueError:
+                pass
+            dir_names.sort()
+            dir_path = Path(dir_path)
             for file_name in sorted(file_names):
-                rel_path = os.path.normpath(os.path.relpath(os.path.join(dir_path, file_name), repo_path))
+                rel_path = (dir_path / file_name).relative_to(repo_path)
                 if rel_path in seen:
                     continue
                 seen.add(rel_path)
                 yield rel_path
 
-    @property
-    def outdated(self):
-        try:
-            with open(self._meta_file, 'r') as raw_meta:
-                meta_data = json.load(raw_meta)
-        except FileNotFoundError:
-            return True
-        else:
-            if meta_data['git_uri'] != self.git_uri or meta_data['branch'] != self.branch:
-                LOGGER.critical('cache %r corrupted by other hook: %r', self, meta_data)
-                raise RuntimeError('config cache %r used for conflicting hooks' % self.cache_path)
-            else:
-                return meta_data['timestamp'] + self.max_age <= time.time()
-
-    def _update_metadata(self):
-        with open(self._meta_file, 'w') as raw_meta:
-            json.dump({'git_uri': self.git_uri, 'branch': self.branch, 'timestamp': time.time()}, raw_meta)
-
-    def refresh(self):
-        if not self.outdated:
+    def _refresh(self):
+        assert self._config_meta is not None
+        self._config_meta["reads"] += 1
+        if self._config_meta["timestamp"] + self.max_age > time.time():
             return
```

Comment:
To improve readability and maintainability you might consider adding a comment here to tell in which cases on can return already.

```suggestion
            return  # early exit as ConfigCache is still valid
```